app/main/views/users.py

```diff
@@ -271,22 +282,42 @@ def export_users_for_framework(framework_slug):
 
     suppliers_with_a_complete_service = frozenset(framework.get_supplier_ids_for_completed_service())
 
-    supplier_id_published_service_count = dict(db.session.query(
-        Service.supplier_id,
-        func.count(Service.id)
-    ).filter(
-        Service.status == 'published',
-        Service.framework_id == framework.id
-    ).group_by(
-        Service.supplier_id
-    ).all())
+    lots = db.session.query(FrameworkLot.lot_id, Lot.slug).join(Lot, FrameworkLot.lot_id == Lot.id).filter(
+        FrameworkLot.framework_id == framework.id
+    ).distinct().all()
+    published_service_count_by_supplier_and_lot = {}
+
+    for lot_id, slug in lots:
+        published_service_count_by_supplier_and_lot[
+            "published_services_count_on_{}_lot".format(slug)
+        ] = dict(db.session.query(
+            Service.supplier_id,
+            func.count(Service.id)
+        ).filter(
+            Service.status == 'published',
+            Service.lot_id == lot_id
+        ).group_by(
+            Service.supplier_id
+        ).group_by(
+            Service.lot_id
+        ).all())
+
+    supplier_id_published_service_count = {}
+    for k, v in published_service_count_by_supplier_and_lot.items():
+        for supplier_id, service_count in v.items():
+            if supplier_id in supplier_id_published_service_count:
+                supplier_id_published_service_count[supplier_id] += service_count
+            else:
+                supplier_id_published_service_count[supplier_id] = service_count
```

Comment:
I tell a lie - neater:
```
from itertools import groupby

lot_slugs_by_id = {lot.id: lot.slug for lot in framework.lots}

service_counts_by_lot_by_supplier = {
    supplier_id: {row[1]: row[2] for row in rows}
    for supplier_id, rows in groupby(
        db.session.query(
            Service.supplier_id,
            Service.lot_id,
            func.count(Service.id)
        ).filter(
            Service.status == 'published',
            Service.framework_id == framework.id
        ).group_by(
            Service.supplier_id,
            Service.lot_id,
        ).all(),
        key=lambda row: row[0],
    )
}
```

Then later:
```
user_rows.append({
    ...
    published_services_count = {
        lot_slugs_by_id[lot_id]: service_counts_by_lot_by_supplier[supplier_id].get(lot_id, 0)
        for lot_id in lot_slugs_by_id.keys()
    }
```

Bam, all done in ~10 additional, declarative lines.

Reply:
Oh actually for backwards compatibility make the latter lines:

```
user_rows.append({
    ...
    "published_services_count": sum(service_counts_by_lot_by_supplier[supplier_id].values())
    "published_services_count_by_lot": {
        lot_slugs_by_id[lot_id]: service_counts_by_lot_by_supplier[supplier_id].get(lot_id, 0)
        for lot_id in lot_slugs_by_id.keys()
    }
})
```

Reply:
p.s. if you want me to explain how this all works and what the difference is, just ask me. Or, if it's obvious, ... don't.